tools/syz-trace2syz/parser/intermediate_types.go

```diff
@@ -0,0 +1,531 @@
+}
+
+func newGroupType(elems []IrType) (typ *GroupType) {
+	return &GroupType{Elems: elems, Len: len(elems)}
+}
+
+// Name implements IrType Name()
+func (a *GroupType) Name() string {
+	return groupTypeName
+}
+
+// String implements IrType String()
+func (a *GroupType) String() string {
+	var buf bytes.Buffer
+
+	buf.WriteString("[")
+	for _, elem := range a.Elems {
+		buf.WriteString(elem.String())
+		buf.WriteString(",")
+	}
+	buf.WriteString("]")
+	return buf.String()
+}
+
+// Field of a struct e.g. name = "Shankara"
+type Field struct {
+	Key string
+	Val IrType
+}
+
+func newField(key string, val IrType) *Field {
+	return &Field{Key: key, Val: val}
+}
+
+// Name implements IrType Name()
+func (f *Field) Name() string {
+	return "Field Type"
+}
+
+func (f *Field) String() string {
+	return f.Val.String()
+}
+
+// Expression represents Ints, Flags, Arithmetic expressions
+type Expression interface {
+	IrType
+	Eval(*prog.Target) uint64
+}
+
+type expressionCommon struct {
+}
+
+// Name implements IrType Name()
+func (e *expressionCommon) Name() string {
+	return exprTypeName
+}
+
+type binOp struct {
+	expressionCommon
+	leftOp  Expression
+	op      operation
+	rightOp Expression
+}
+
+func newBinop(leftOperand, rightOperand IrType, Op operation) *binOp {
+	return &binOp{leftOp: leftOperand.(Expression), rightOp: rightOperand.(Expression), op: Op}
+}
+
+// Eval implements Expression's Eval()
+func (b *binOp) Eval(target *prog.Target) uint64 {
+	op1Eval := b.leftOp.Eval(target)
+	op2Eval := b.rightOp.Eval(target)
+	switch b.op {
+	case andOp:
+		return op1Eval & op2Eval
+	case orOp:
+		return op1Eval | op2Eval
+	case xorOp:
+		return op1Eval ^ op2Eval
+	case lshiftOp:
+		return op1Eval << op2Eval
+	case rshiftOp:
+		return op1Eval >> op2Eval
+	case timesOp:
+		return op1Eval * op2Eval
+	case minusOp:
+		return op1Eval - op2Eval
+	case plusOp:
+		return op1Eval + op2Eval
+	default:
+		log.Fatalf("Unable to handle op: %d", b.op)
+		return 0
+	}
+}
+
+// String implements IrType String()
+func (b *binOp) String() string {
+	return fmt.Sprintf("op1: %s op2: %s, operand: %v\n", b.leftOp.String(), b.rightOp.String(), b.op)
+}
+
+type unOp struct {
+	expressionCommon
+	op      operation
+	operand Expression
+}
+
+func newUnop(operand IrType, op operation) *unOp {
+	return &unOp{op: op, operand: operand.(Expression)}
+}
+
+// Eval implements Expression's Eval()
+func (u *unOp) Eval(target *prog.Target) uint64 {
+	opEval := u.operand.Eval(target)
+	switch u.op {
+	case onescompOp:
+		return ^opEval
+	case negOp:
+		return -opEval
+	default:
+		log.Fatalf("Unsupported Unop Op: %d", u.op)
+	}
+	return 0
+}
+
+// String implements IrType String()
+func (u *unOp) String() string {
+	return fmt.Sprintf("op1: %v operand: %v\n", u.operand, u.op)
+}
+
+// DynamicType represents cases where strace shows the before system call and after
+// E.g. <... getsockname resumed> {sa_family=AF_INET6,..., sin6_scope_id=0}, [112->28]
+type DynamicType struct {
```

Comment:
AfterCall is never used. Why do we need DynamicType at all? Can we remove it and use BeforeCall expressions instead of it?

Reply:
Yes, it would be preferable. Here are so many things and indirection that I still can't get overall picture, I am just looking at individual things and trying to understand "what is this?", "why do we need this?", "where we use it?", "what's the difference between BeforeCall and AfterCall and where we use one or another?". And in the end it turns out that we just don't need it, so it's not actually possible to answer all these questions.
Writing code is easy and cheap. It's reviewing it, testing, making sure it's correct and maintaining it over time that's expensive. It's always better to start small, and add stuff if and when it's actually necessary, rather then go in the opposite direction.